### ISSUE TITLE
Status log duplicates after Order Placed & Paid 

### ISSUE THREAD

**Issue by Web10-Joris** [damcclean/simple-commerce#1113]
### Description

As partially mentioned in  #1110.

I've got a weird problem with order status logs in my setup. It's creating duplicate entries and I can't figure out why.
Here's what's happening:

1. When an order is placed and paid, it's creating multiple entries in the status log table. Some timestamps are even out of order.

    | id | order_id | status | timestamp | data |
    |----|----------|--------|-----------|------|
    | 8  | 3        | placed | 2024-07-29 17:01:17 | [] |
    | 9  | 3        | placed | 2024-07-29 15:01:17 | [] |
    | 10 | 3        | paid   | 2024-07-29 17:01:17 | [] |
    | 11 | 3        | placed | 2024-07-29 13:01:17 | [] |
    | 12 | 3        | paid   | 2024-07-29 15:01:17 | [] |

3. The order->data->status_log in the orders table is also showing these duplicate entries. _Is it correct that this is still written to the data->status_log after the update to have a separate status_log table?_ 
    ```
    {
	    "site": "default",
	    "mollie": {
		    "id": "tr_2wskwM5dnK",
		    "checkout_url": "xxx"
	    },
	    "status_log": [
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722265277 // -> 2024-07-29 17:01:17
		    },
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722258077 // -> 2024-07-29 15:01:17
		    },
		    {
			    "data": [],
			    "status": "paid",
			    "timestamp": 1722265277 // -> 2024-07-29 17:01:17
		    }
	    ],
	    "shipping_tax": {
		    "rate": 21,
		    "amount": 174,
		    "price_includes_tax": true
	    },
	    "shipping_method": "sendcloud"
    }
    ```

4. When I update an order to "Dispatched", it adds even more duplicate entries to both the status log table and the order->data->status_log.

    | id | order_id | status     | timestamp           | data                                        |
    |----|----------|------------|---------------------|---------------------------------------------|
    | _8_  | 3        | placed     | 2024-07-29 17:01:17 | []                                          |
    | _9_  | 3        | placed     | 2024-07-29 15:01:17 | []                                          |
    | _10_ | 3        | paid       | 2024-07-29 17:01:17 | []                                          |
    | _11_ | 3        | placed     | 2024-07-29 13:01:17 | []                                          |
    | _12_ | 3        | paid       | 2024-07-29 15:01:17 | []                                          |
    | **31** | 3        | placed     | 2024-07-29 11:01:17 | []                                          |
    | **32** | 3        | paid       | 2024-07-29 13:01:17 | []                                          |
    | **33** | 3        | dispatched | 2024-07-29 18:07:54 | {"user": 1, "reason": "Test status change"} |
    | **34** | 3        | placed     | 2024-07-29 09:01:17 | []                                          |
    | **35** | 3        | paid       | 2024-07-29 11:01:17 | []                                          |
    | **36** | 3        | dispatched | 2024-07-29 16:07:54 | {"user": 1, "reason": "Test status change"} |

    ```
    "status_log": [
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722265277 // -> 2024-07-29 17:01:17
		    },
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722258077 // -> 2024-07-29 15:01:17
		    },
		    {
			    "data": [],
			    "status": "paid",
			    "timestamp": 1722265277 // -> 2024-07-29 17:01:17
		    },
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722250877 // -> 2024-07-29 13:01:17
		    },
		    {
			    "data": [],
			    "status": "paid",
			    "timestamp": 1722258077 // -> 2024-07-29 15:01:17
		    },
		    {
			    "data": [],
			    "status": "placed",
			    "timestamp": 1722243677 // -> 2024-07-29 11:01:17
		    },
		    {
			    "data": [],
			    "status": "paid",
			    "timestamp": 1722250877 // -> 2024-07-29 13:01:17
		    },
		    {
			    "data": {
				    "user": 1,
				    "reason": "Test status change"
			    },
			    "status": "dispatched",
			    "timestamp": 1722269274 // -> 2024-07-29 18:07:54
		    }
	    ],
    ```
I'm not using any custom code that should affect this, except for a small tweak to filter out cart orders. No other unusual adjustments related to status log or orders:
```
    public function scopeRunwayListing(Builder $query): Builder
    {
        return $query
            ->where('order_status', '!=', OrderStatus::Cart)  // Exclude 'Cart' orders
            ->leftJoin('status_log', function ($join) {
                $join->on('status_log.order_id', '=', 'orders.id')
                    ->where('status_log.status', 'placed');
            })
            ->orderByRaw('COALESCE(status_log.timestamp, orders.created_at) DESC')  // Order by the most recent timestamp
            ->select('orders.*');  // Select all columns from 'orders'
    }
```

The checkout flow is normal, with a redirect to Mollie for payment. The checkout page is a one-pager that updates the order using AJAX when Country / Address is filled in to recalculate with selected the shipping_method and tax-zone/tax-rate. 
Any ideas on what might be causing this or where I should look to debug further? 

### Steps to reproduce

1. Use database orders
2. Placed order locally with ngrok for Mollie to reach 
3. Results in multiple status logs 
4. Update status to Dispatched to see even more duplicated logs 

### Environment

Application Name: XXX
Laravel Version: 11.18.1
PHP Version: 8.3.7
Composer Version: 2.5.5
Environment: local
Debug Mode: ENABLED
URL: localhost/
Maintenance Mode: OFF
Timezone: Europe/Amsterdam
Locale: en

Cache
Config: NOT CACHED
Events: NOT CACHED
Routes: NOT CACHED
Views: CACHED

Drivers
Broadcasting: null
Cache: file
Database: mysql
Logs: stack / single
Mail: smtp
Queue: database
Session: file

Simple Commerce
Currencies: EUR
Gateways: Mollie
Repository: Customer: DuncanMcClean\SimpleCommerce\Customers\EloquentCustomerRepository
Repository: Order: DuncanMcClean\SimpleCommerce\Orders\EloquentOrderRepository
Repository: Product: DuncanMcClean\SimpleCommerce\Products\EntryProductRepository
Shipping Methods: shipping_methods.collect.title, shipping_methods.sendcloud.title, shipping_methods.brenger.title, shipping_methods.mtc.title, shipping_methods.quicargo.title
Tax Engine: DuncanMcClean\SimpleCommerce\Tax\Standard\TaxEngine

Statamic
Addons: 4
Sites: 1
Stache Watcher: Disabled
Static Caching: Disabled
Version: 5.17.0 PRO

Statamic Addons
aerni/advanced-seo: 2.8.0
aerni/font-awesome: 2.1.1
duncanmcclean/simple-commerce: 7.4.0
statamic-rad-pack/runway: 7.7.2

**Comment by duncanmcclean** [comment]
> When an order is placed and paid, it's creating multiple entries in the status log table. Some timestamps are even out of order.

Technically, it *is* possible for the same event to be logged in the status log multiple times (in case a payment doesn't go through for whatever reason, it can be moved back to "Unpaid" with a log of what happened). 

Whenever SC needs to pull a timestamp for an event, it'll find the latest instance of that event in the status log and use that.

I'm not quite sure why they're out-of-order. It should just be merging in the existing status log array with the new one (or be adding rows, in the case of database orders).

> The order->data->status_log in the orders table is also showing these duplicate entries. Is it correct that this is still written to the data->status_log after the update to have a separate status_log table?

No, the `status_log` key shouldn't exist in the `data` object when using storing orders in the database, they should only be logged to the `status_log` table.

As part of the status log migration, after creating the table and moving over existing data, it should have removed the `status_log` key from the `data` column: https://github.com/duncanmcclean/simple-commerce/blob/7.x/src/Console/Commands/stubs/create_status_log_table.php#L39

> When I update an order to "Dispatched", it adds even more duplicate entries to both the status log table and the order->data->status_log.

Hmm, it definetly shouldn't be adding duplicates entries for the same "event" (eg. if you change the status to "Dispatched", it should only add a single row, not multiple).

Are you able to replicate this issue on a fresh site?

---

**As a side note:** I'm currently working on a new version of Simple Commerce, which'll do away with the status log and simplify things back to how they used to be. (I have no ETA on this though, but wanted to let you know I don't like the status log system either 😬 )

**Comment by Web10-Joris** [comment]
Thanks for the extensive answer Duncan! 

I ran a bit of testing on a fresh install (SC Starter kit with database orders enabled). Here the status log items also still show up in the `order->data->status_log` along with the items in the `status_log` table. 

I found out that when I then change the `APP_LOCALE` from `UTC` to `Europe/Amsterdam` (on the fresh install) I get the strange behavior with duplicated events and inconsistent sequences.
